snowflakedb/snowflake-cli
# Workspaces application package - drop action and command

### Pre-review checklist
   * [x] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * I've described my changes in the release notes.

### Changes description

* Adding `snow ws drop` command
* Implementing `ApplicationPackageEntity.action_drop()` and wiring `NativeAppTeardownProcessor.drop_package()` to it
* Extracting `drop_generic_object()` to a stateless utility
* Slightly changing the distribution warning message to not be "snow app" specific:
```diff
- Continuing to execute `snow app teardown` on application package app_pkg with distribution 'external'.
+ Dropping application package app_pkg with distribution 'external'.
```

### Manual test

Verifying `--force` is required when a version exists:
```
$ snow ws deploy --entity-id=pkg
$ snow app version create v1

$ snow ws drop --entity-id=pkg
╭─ Error ───────────────────────────────────────────────────────────────────────────╮
│ Application package could not be dropped as it has versions associated with it.   │
│ Drop versions first, or use --force to override.                                  │
╰───────────────────────────────────────────────────────────────────────────────────╯

$ snow ws drop --entity-id=pkg --force
Dropping application package pdfv2_pkg_gbloom now.
Dropped application package pdfv2_pkg_gbloom successfully.
```

Verifying user is prompted to confirm drop when no special comment:
```
$ snow ws deploy --entity-id=pkg
$ snow sql -q "alter application package pdfv2_pkg_gbloom set comment='foo'"

$ snow ws drop --entity-id=pkg
Application package pdfv2_pkg_gbloom was not created by Snowflake CLI.
Application package details:
Name: pdfv2_pkg_gbloom
Created on: 2024-08-29 12:46:05.972000-07:00
Distribution: internal
Owner: ACCOUNTADMIN
Comment: foo
Are you sure you want to drop it?
 [y/N]: N
Did not drop application package pdfv2_pkg_gbloom.

$ snow ws drop --entity-id=pkg
Application package pdfv2_pkg_gbloom was not created by Snowflake CLI.
Application package details:
Name: pdfv2_pkg_gbloom
Created on: 2024-08-29 12:46:05.972000-07:00
Distribution: internal
Owner: ACCOUNTADMIN
Comment: foo
Are you sure you want to drop it?
 [y/N]: y
Dropping application package pdfv2_pkg_gbloom now.
Dropped application package pdfv2_pkg_gbloom successfully.
```
